### PR TITLE
Delete pods assigned to not existing nodes.

### LINKED ISSUE #71349: DaemonSet Controller doesn't delete orphaned pods
**What happened**:
DaemonSet doesn't delete pods that:
- weren't scheduled yet,
- node was deleted so pod can't be scheduled.
**What you expected to happen**:
Pod should be deleted when node is removed.
**How to reproduce it (as minimally and precisely as possible)**:
I encountered this problem when doing tests with cluster with 1000 nodes and 30000 pods running. During tests I was deleting and adding nodes using GCE API without draining them. I reproduced problem on much smaller cluster but it requires stopping K8s scheduler.

Scenario:
1. Create cluster with 2 nodes.
2. Wait till fluentd-gcp pods are running on nodes.
3. Stop K8s scheduler.
4. Delete fluentd-gcp pods.
5. Wait till daemonset controller creates pending pods.
6. Delete one node.
7. Start scheduler.
8. One of the fluentd-gcp pods will be pending and scheduler won't be able to schedule.

**Anything else we need to know?**:

**Environment**: GKE
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.7", GitCommit:"0c38c362511b20a098d7cd855f1314dad92c2780", GitTreeState:"clean", BuildDate:"2018-08-20T10:09:03Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"12+", GitVersion:"v1.12.2-gke.3", GitCommit:"3bdb4f77629a001276ae061e68aff4bac147fbc5", GitTreeState:"clean", BuildDate:"2018-11-13T22:54:51Z", GoVersion:"go1.10.4b4", Compiler:"gc", Platform:"linux/amd64"}
```

/kind bug

### PR DESCRIPTION
**Pull request by krzysztof-jastrzebski**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
PR adds deleting pods created by DaemonSet assigned to not existing nodes.
**Which issue(s) this PR fixes**:
Fixes #71349

**Does this PR introduce a user-facing change?**:

```release-note
Adds deleting pods created by DaemonSet assigned to not existing nodes.
```

### CODE REVIEW

File: pkg/controller/daemon/daemon_controller.go [hunk 1]
```diff
@@ -1524,3 +1527,21 @@ func (o podByCreationTimestampAndPhase) Less(i, j int) bool {
 func failedPodsBackoffKey(ds *apps.DaemonSet, nodeName string) string {
 	return fmt.Sprintf("%s/%d/%s", ds.UID, ds.Status.ObservedGeneration, nodeName)
 }
+
+// getPodsWithoutNode returns list of pods assigned to not existing nodes.
+func getPodsWithoutNode(
+	runningNodesList []*v1.Node, nodeToDaemonPods map[string][]*v1.Pod) []string {
+	var results []string
+	isNodeRunning := make(map[string]bool)
+	for _, node := range runningNodesList {
```
Comment (k82cn) [inline pkg/controller/daemon/daemon_controller.go]: why not dependent on `PodGCController` ?

xref https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/podgc/gc_controller.go#L143

File: pkg/controller/daemon/daemon_controller.go [hunk 2]
```diff
@@ -1524,3 +1527,21 @@ func (o podByCreationTimestampAndPhase) Less(i, j int) bool {
 func failedPodsBackoffKey(ds *apps.DaemonSet, nodeName string) string {
 	return fmt.Sprintf("%s/%d/%s", ds.UID, ds.Status.ObservedGeneration, nodeName)
 }
+
+// getPodsWithoutNode returns list of pods assigned to not existing nodes.
+func getPodsWithoutNode(
+	runningNodesList []*v1.Node, nodeToDaemonPods map[string][]*v1.Pod) []string {
```
Comment (k82cn) [inline pkg/controller/daemon/daemon_controller.go]: move to a single line :)

**Comment by k82cn** [comment]
LGTM ; let's get it merged when CI's happy :)

**Comment by mikedanese** [comment]
isn't this smoothing that the node controller should handle in general? was the problem that they were critical pods?

**Comment by seh** [comment]
These pods weren't assigned to nodes by the scheduler yet; they were created with affinity that will only allow them to be placed on a single node—in each case, a node that no longer exists.

This happens with pods regardless of whether they're critical or not. We see it happen with pods that need to wait until the node is ready, where the node never became ready before it got deleted.

**Comment by k82cn** [comment]
> isn't this smoothing that the node controller should handle in general? was the problem that they were critical pods?

`PodGC` will check pod's node name to clean them up. After `ScheduleDaemonPod` (schedule daemonset pods by scheduler), `DemonSetController` will only append node affinity with target node and let scheduler to do binding. So it's better to handle it in `DaemonSetController`.

BTW, @krzysztof-jastrzebski , would you help to add feature-gate checking there? `ScheduleDaemonPod ` is still beta now; if anyone disable it, we don-t need to check all-nodes and all-pods :)

**Pull request merged**